_posts/pl/2021-11-23-json-vs-jsonb-postgresql.md

```diff
@@ -0,0 +1,97 @@
+---
+layout:    post
+title:     "Czy wiesz czym różnią się json i jsonb w PostgreSQL"
+date:      2021-12-01 8:00:00 +0100
+published: true
+didyouknow: true
+lang: pl
+author:    kkur
+image:     /assets/img/posts/2021-11-23-json-vs-jsonb-postgresql/json.png
+tags:
+- PostgreSQL
+---
+
+# Podstawowe różnice
+
+Podstawową różnicą pomiędzy json a jsonb jest sposób ich przechowywania. Typ json jest przechowywany jako tekst, natomiast typ jsonb jest przechowywany w postaci binarnej. Sposób przechowywania ma wpływ na kilka czynników:
+
+* zapis danych w postaci jsonb trwa dłużej,
+* operacje wykonywane na typie jsonb trwają krócej, gdyż nie trzeba ich dodatkowo parsować.
+
+Ze sposobu przechowywania wynikają kolejne różnice. Zobaczmy poniższy przykład:
+
+```sql
+
+SELECT '{"c":0, "a":2,"a":1}'::json, '{"c":0, "a":2,"a":1}'::jsonb;
+```
+
+``` 
+          json          |        jsonb
+------------------------+---------------------
+ {"c":0,   "a":2,"a":1} | {"a": 1, "c": 0}
+```
+* Json zachowuje formatowanie - dlatego w przykładzie widać spacje przed kluczem a, natomiast jsonb nie zachowuje formatowania - usuwa whitespace'y.
+* Json pozwala dodawać powielone klucze (w przykładzie widać, że klucz "a" występuje 2 razy), natomiast jsonb zachowuje tylko ostatnią wartość.
+* Json zachowuje kolejność kluczy, natomiast jsonb przechowuje posortowane alfabetycznie klucze.
+* Jsonb przechowuje obiekty w postaci binarnej, natomiast json przechowuje dane jako tekst.
+* Jsonb pozwala zakładać indexy na kluczach.
+
+# Operatory wspierane tylko przez jsonb
+W celu zobrazowania dodatkowych operatorów dla typu jsonb, zacznijmy od utworzenia prostej tabeli, którą wykorzystamy w przykładach
+
+```sql
+CREATE TABLE orders (
+    id serial NOT NULL PRIMARY KEY,
+    data jsonb NOT NULL
+);
+```
+
+oraz zainicjujmy ją prostymi danymi w formacie json.
+
+```sql
+
+INSERT INTO orders (data) VALUES ('{ "customer": "Anna Kowalczyk", "items": {"product": "Pieluchy","qty": 24} }'),
+    ('{ "customer": "Zofia Wiśniewska", "items": {"product": "Zabawka samochód","qty": 1} }'),
+    ('{ "customer": "Wojciech Nowak", "items": {"product": "Klocki LEGO","qty": "1"}, "regular_customer": true}'),
+    ('{ "customer": "Jan Kowalski", "items": {"product": "Piwo","qty": 6}}');
+```
+Operator `@>` pozwala na sprawdzenie, czy json zawiera na tym poziomie pole wraz z podaną wartością.
+
+```sql
+
+SELECT (data -> 'items') @> '{"qty": 6}' from orders;
+  
+contains
+----------
+f
+f
+f
+```
+
+Operator `?` pozwala na sprawdzenie, czy json posiada określone pole. Przydatne, jeśli jsony nie trzymają się restrykcyjnie określonego schematu i obiekty mogą posiadać różne zestawy pól.
+```sql
+SELECT data ? 'regular_customer' from orders;
+  
+field_exists
+----------
+f
+f
+t
+f
+```
+
+Poza tym typ jsonb wspiera kilka operatorów, które pozwalają na skuteczną modyfikację jsonów trzymanych w bazie:
```

Comment:
Z uwagi na kolejne punkty, to zdanie bym zakończyła kropką, a kolejne punkty "pozwala...." dała od punktów/ kropek. 
__Dodatkowo typ jsonb wspiera kilka operatorów, które pozwalają na skuteczną modyfikację jsonów trzymanych w bazie._

Reply:
done